File: PorpoiseOfClapping/Assets/Tiny/PorpoiseOfClapping/Scripts/OnTapPhaseIncrementerSystem.ts
namespace game {
    export class OnTapPhaseIncrementerSystem extends ut.ComponentSystem {
        OnUpdate():void {
            let phaseConfig = this.world.getConfigData(game.PhaseConfig);
            let deltaTime:number = this.scheduler.deltaTime();
            if (phaseConfig.changed) {
                phaseConfig.changed = false;
            }
            if (phaseConfig.timeElapsed == 0 && phaseConfig.phase == 0) {
                phaseConfig.changed = true;
            }
            phaseConfig.timeElapsed += deltaTime;

            let mouseDown:boolean = ut.Runtime.Input.getMouseButtonDown(0);
            if (!mouseDown) {
                this.world.setConfigData(phaseConfig);
                return;
            }

            console.log("OnTapPhaseIncrementerSystem: mouseDown");
            if (phaseConfig.phase > 0 && phaseConfig.timeElapsed < phaseConfig.minDuration) {
                this.world.setConfigData(phaseConfig);
                return;
            }

            phaseConfig.phase++;
            console.log("OnTapPhaseIncrementerSystem: changed phase=" + phaseConfig.phase +
                " time elapsed=" + phaseConfig.timeElapsed);
            phaseConfig.timeElapsed = 0;
            phaseConfig.changed = true;
            this.world.setConfigData(phaseConfig);
        }
    }
}

File: PorpoiseOfClapping/Assets/Scripts/OnTapPhaseIncrementerSystem.cs
using Unity.Burst;
using Unity.Collections;
using Unity.Entities;
using Unity.Jobs;
using Unity.Mathematics;
using Unity.Transforms;
using UnityEngine;

namespace Game
{
    public class OnTapPhaseIncrementerSystem : JobComponentSystem
    {
        [BurstCompile]
        struct OnTapPhaseIncrementerJob : IJobProcessComponentData<PhaseConfig>
        {
            [ReadOnly] public float deltaTime;
            [ReadOnly] public bool mouseDown;

            public void Execute(ref PhaseConfig phaseConfig)
            {
                if (phaseConfig.changed)
                {
                    phaseConfig.changed = false;
                }
                if (phaseConfig.timeElapsed == 0 && phaseConfig.phase == 0)
                {
                    phaseConfig.changed = true;
                }
                phaseConfig.timeElapsed += deltaTime;

                if (!mouseDown)
                {
                    return;
                }

                if (phaseConfig.phase > 0 && phaseConfig.timeElapsed < phaseConfig.minDuration)
                {
                    return;
                }

                phaseConfig.phase++;
                if (phaseConfig.phase > phaseConfig.maxPhase)
                {
                    phaseConfig.phase = 0;
                }
                phaseConfig.timeElapsed = 0;
                phaseConfig.changed = true;
            }
        }

        protected override JobHandle OnUpdate(JobHandle inputDependencies)
        {
            var job = new OnTapPhaseIncrementerJob()
            {
                deltaTime = Time.deltaTime,
                mouseDown = Input.GetMouseButtonDown(0)
            };

            return job.Schedule(this, inputDependencies);
        }

    }
}

